### PR TITLE
More shape checking for ConvNd

### LINKED ISSUE #3039: torch.nn.functional.conv2d seg faults on a wrongly shaped kernel
This snippet
```
from torch import Tensor
from torch.autograd import Variable
from torch.nn.functional import conv2d
k = Variable(Tensor([[1]]))
x = Variable(Tensor(1, 1, 1, 1))
conv2d(x, k)
```
seg faults on my install.

### PR DESCRIPTION
**Pull request by ssnl**
Provides better error handle and message. Fixes #3039 .

### CODE REVIEW

**Comment by soumith** [comment]
failing basic contbuild (not sure how it compiled for you locally)

**Comment by SsnL** [comment]
Whoops my bad. Should be compilable now.

**Comment by killeent** [comment]
At some point we need to figure out how to unify `Conv`, including where the shape-checking logic should rely.

**Comment by SsnL** [comment]
There's something wrong about my understanding of bias. I'll fix these errors.

**Comment by SsnL** [comment]
Didn't consider the transpose case :) Now it's fixed.

File: torch/csrc/autograd/functions/convolution.cpp [hunk 1]
```diff
@@ -156,26 +156,57 @@ static auto view3d(const at::Tensor& tensor) -> at::Tensor {
...
 
 static void check_input_shape_forward(const at::Tensor& input,
-				      const at::Tensor& weight,
+				      const at::Tensor& weight, const at::Tensor& bias,
 				      int64_t groups, bool transposed) {
+  int k = input.ndimension();
+
+  if (weight.ndimension() != k) {
+      std::stringstream ss;
+      ss << "Expected " << k << "-dimensional weight for " << k
```
Comment (soumith) [inline torch/csrc/autograd/functions/convolution.cpp]: it's the other way around: "Expected K-dimensional input for K-dimensional weight, but got input of size..."

**Review by soumith — CHANGES_REQUESTED** [review]

**Review by soumith — APPROVED** [review]

**Pull request merged**